Iceberg with data migrations

bashtanov opened the pull request:
https://redpandadata.atlassian.net/browse/CORE-8439
- Add a test for iceberg to read from table whose topic was deleted
- Fix minor data migration test issues
- Add a test to run iceberg translation for topics unmounted and then, optionally, mounted
- For recovered and mounted topics, make Redpanda preserve most topic properties including iceberg ones (fixes https://redpandadata.atlassian.net/browse/CORE-563)
- When unmounting make sure all messages are translated for iceberg

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
### Features
* Make Iceberg and topic mount/unmount work well together
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
utils/notification_list: collect return values

Allows to use it for subscriptions where feedback from a called function
is necessary, such as a future or an error code.
All functions are supposed to return the same type.

bashtanov committed:
r/offset_monitor: make templated so it can be used for kafka offsets

Make offset_monitor more universal so that it can be used for different
data types.

bashtanov committed:
c/topic_recovery_service: copy all topic properties, override specific

This is mostly to preserve iceberg properties, but also to make sure any
newly introduced topic properties are preserved by default.

bashtanov committed:
c/migrations/backend: copy all topic properties, override specific ones

This is mostly to preserve iceberg properties, but also to make sure any
newly introduced topic properties are preserved by default.

bashtanov committed:
c/archival/ntp_archiver_service: hold gate while waiting for flush

to prevent archiver shutdown while waiting

bashtanov committed:
dl/translation/stm: implement waiting for translation of specific offset

bashtanov committed:
model/fundamental: fix typo in comment

bashtanov committed:
c/topic_frontend: fix formatting

bashtanov committed:
c/partition: to dispatch async flush actions to components

Also create and subscribe one of these actions: flush data to cloud.

bashtanov committed:
c/partition: add comment

bashtanov committed:
dl/translation/partition_translator: subscribe as a partition flusher

Wait for the offset to be translated when asked by partition to "flush".

bashtanov committed:
c/migrations/worker: get offset when blocking writes and use it on flush

When blocking writes collect the offset of the blocking message.
Then use it to dispatch all-components flush through partition
(leading to cloud storage flush that ignores the offset parameter and
datalake translator that waits for the correspondent kafka offset)

bashtanov committed:
tests/dl/verifier: improve logging

add log lines, fix typos

bashtanov committed:
tests/services/connect/stop: option to make sure stream NOT finished

for to make functionality is tested while topic is being actively used

bashtanov committed:
tests/datalake/e2e: table can be queried when topic is deleted

Check that with redpanda.iceberg.delete=false old table data remains
available even before we recreate the topic.

bashtanov committed:
tests/dl/verifier: add facility to wait until first message via iceberg

if we unmount the topic before this table may lack metadata

bashtanov committed:
tests/dl/verifier: introduce mode for no communication with the topic

Introduce "offline mode" that cuts all ties to the topic in Redpanda
cluster. It carries on querying the query engine and verifying results
using info cached before going into offline mode.

bashtanov committed:
tests/dl/simple_connect_test: refactor in prep to add un/mount tests

bashtanov committed:
tests/dl/simple_connect_test/rpconnect: configure no of messages

Make it possible to configure the number of messages produced by stream

bashtanov committed:
tests/dl/simple_connect_test: test with data migrations

Add scenarios:
1) On unmount all messages that made their way to the topic eventually
become available via query engine
2) Upon remount and further produce both old and new messages are in the
topic and in the table

bashtanov committed:
tests/migrations: switch to flaky admin before disruption

And switch back to normal admin after disruptions are over.

bashtanov committed:
tests/migrations: separate migration utility functions into a mixin

bashtanov committed:
tests/migrations: correct test_listing_inexistent_migration

bashtanov commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e48fd6d7-d066-4b95-8496-e0179699cbdf-->
<!--TEST RESULTS-->
<!--2fddceaa-c55b-4c23-aee3-c9a4d38f6c20-->
<!--TEST RESULTS-->
<!--71e42f50-1a78-4d47-88e5-805286efa748-->
<!--TEST RESULTS-->
<!--04b1d11a-4725-4352-a811-d30545f04530-->
<!--TEST RESULTS-->
<!--3eb4d85d-b4e0-43d6-bdc0-47859cb72a0e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60655">build#60655</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| idempotency_tests_rpunit.idempotency_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60655#01946058-a634-40d0-9eaa-a36681179d0c | FLAKY | 1/2 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60655#019460b1-911e-438a-9908-47112e140ed0 | FLAKY | 2/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60858">build#60858</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| idempotency_tests_rpunit.idempotency_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60858#0194702b-4eab-49b6-a88e-86bc997a49fa | FLAKY | 1/2 |
| rptest.tests.datalake.simple_connect_test.RedpandaConnectIcebergTest.test_translating_avro_serialized_records.cloud_storage_type=CloudStorageType.S3.scenario=remount | ducktape | https://buildkite.com/redpanda/redpanda/builds/60858#01947072-773f-429b-924e-88e414d4e4b8 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61039">build#61039</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61039#01948df4-24e0-4a1b-adc7-427bf0a8eec8 | FLAKY | 1/2 |
| rptest.tests.topic_creation_test.TopicRecreateTest.test_topic_recreation_while_producing.workload=IDEMPOTENT.cleanup_policy=delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/61039#01948e11-1c37-43dc-9d36-c64694dbeea3 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61291">build#61291</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61291#0194ad0b-4239-400c-99f3-ec7ba3ebc4e8 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61291#0194ad0b-4239-4508-9eb4-142a5a4211f1 | FLAKY | 1/2 |
| storage_single_thread_rpunit.storage_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61291#0194acab-90cb-41b8-8d84-aa2c0d1c0c2e | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61361">build#61361</a></summary>


| test_id                                                                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61361#0194b306-2928-4576-bc99-42bbc006d21c | FLAKY         | 1/2      |
| rptest.tests.enterprise_features_license_test.EnterpriseFeaturesTest.test_enable_features.feature=Feature.oidc.install_license=True.disable_trial=True                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61361#0194b30b-c7e9-4daf-8a19-487158760573 | FLAKY         | 1/2      |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61361#0194b30b-c7e8-4103-bbfb-cf32306b71a9 | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61361#0194b30b-c7e7-4f8b-96d2-fb381349c39f | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61361#0194b30b-c7e7-4f8b-96d2-fb381349c39f | FLAKY         | 1/2      |
</details>
<!--/body-->

mmaslankaprv commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -79,6 +79,12 @@ def __init__(self,
+        self._offline_mode_requested = threading.Event()
+        self._consumer_stopped = threading.Event()
+        self._offline_mode_established = False
+        self._consumer_positions = None  # set iff in offline mode
```
nit: typo

bashtanov replied:
https://en.wikipedia.org/wiki/If_and_only_if

mmaslankaprv commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -103,6 +109,10 @@ def create_consumer(self):
         return c
 
     def update_and_get_fetch_positions(self):
+        if self._offline_mode_established:
```
Can we add a comment on top of the class describing modes of operation ?

mmaslankaprv commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -3001,6 +3001,8 @@ flush_result ntp_archiver::flush() {
 }
 
 ss::future<wait_result> ntp_archiver::wait(model::offset o) {
+    ss::gate::holder holder(_gate);
```
`flush` may take some time, are we certain that when partition is stopped the `flush` will be interrupted ?

bashtanov replied:
The only `co_await` in this function is `co_await _flush_cond.wait()`. When the gate is closed `_flush_cond` is already broken, see [there](https://github.com/redpanda-data/redpanda/pull/24780/files/f677cc34f585e0ee28b93c4a2bc9e7ab94b085d2..86f894b90ec752d926d200cfc99eacd0d11d263b#diff-1497192be88ee6c90154877e16f51c3b5d241969b5394d0b7598f6d7e30f09c0R1205-L1206) (it took GH frontend 30 seconds to load this link in my browser!)

mmaslankaprv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -107,12 +115,30 @@ class coordinator_table_creator : public table_creator {
 constexpr ::model::timeout_clock::duration wait_timeout = 5s;
 
-partition_translator::~partition_translator() = default;
+partition_translator::~partition_translator() {
+    _partition->unregister_flush_hook(_partition_flush_subscription);
+};
```
should that be unregistered in `stop()` ?

bashtanov replied:
it does not seem to make any difference, as the hook function doesn't even need the object to remain live after its scheduling point. But I'll move it to `stop()` to make things neat.

mmaslankaprv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -24,12 +25,19 @@
+#include <seastar/core/future.hh>
 #include <seastar/coroutine/as_future.hh>
 
+#include <functional>
```
please remove unused includes

bashtanov replied:
Which include do you think is unused?

`<functional>` is used [here](https://github.com/redpanda-data/redpanda/pull/24780/commits/7d90b87b99019ad32b74cfb1af5f11a339ba9682#diff-617058a8220c70fdc909b169e594470f7b81ed70497f71c95f421c9da374aee8R160).

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60858-->
<!--c634f5c1-ef32-42de-bb74-e6bc175bceb9-->
<!--title-->
## Retry command for [Build#60858](https://buildkite.com/redpanda/redpanda/builds/60858)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/simple_connect_test.py::RedpandaConnectIcebergTest.test_translating_avro_serialized_records@{"cloud_storage_type":1,"scenario":"remount"}
```
<!--/body-->

bashtanov commented:
Meh. It does not fail when I run it locally with repeat.

bashtanov commented:
I increased timeout in the test, as it coordinator loop, as one last use of the long one may be in progress while we are waiting. Also added some logging and removed dead code. Please re-review.

bashtanov commented:
/dt

mmaslankaprv commented on src/v/cluster/data_migration_worker.cc:
```diff
@@ -248,16 +249,21 @@ ss::future<errc> worker::do_work(
-        co_return co_await flush(partition);
+        auto block_offset = block_res.value();
+
+        auto deadline = model::timeout_clock::now() + 5s;
```
It seems to me that this timeout value is very low considering the possible amount of work is there to be done. Should we consider making it larger and configurable ?

bashtanov replied:
It will be retried by migration reconciliation, including when controller leadership changes. Since partition is blocked already translation backlog will eventually get processed. Translator flush doesn't do any active work, it just waits for certain offset to be translated. The only problem is cloud storage flush will be invoked every time. @WillemKauf @andrwng if a partition does not receive any further writes is it much overhead to flush its cloud data every few seconds?

mmaslankaprv replied:
With this call we are waiting for the translator to actually execute the translation work, and this may take a while depending on the translation gap. If all the calls are idempotent then this is great, it will just be retried and eventually succeed

bashtanov commented:
/dt
